src/tools/rbd_mirror/image_map/Policy.cc

```diff
@@ -0,0 +1,200 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/debug.h"
+#include "common/errno.h"
+
+#include "Policy.h"
+#include "librbd/Utils.h"
+#include "LoadRequest.h"
+#include "UpdateRequest.h"
+#include "RemoveRequest.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rbd_mirror
+#undef dout_prefix
+#define dout_prefix *_dout << "rbd::mirror::image_map::Policy: " << this \
+                           << " " << __func__
+
+namespace rbd {
+namespace mirror {
+
+using rbd::mirror::ImageIds;
+
+namespace image_map {
+
+using librbd::util::unique_lock_name;
+using librbd::util::create_context_callback;
+
+template <typename I>
+const std::string Policy<I>::UNMAPPED_INSTANCE_ID("");
+
+template <typename I>
+Policy<I>::Policy(librados::IoCtx &ioctx)
+  : m_lock(unique_lock_name("rbd::mirror::image_map::Policy::m_lock", this)),
+    m_ioctx(ioctx) {
+}
+
+template <typename I>
+std::string Policy<I>::lookup(const std::string &global_image_id) {
+  assert(m_lock.is_locked());
+
+  for (auto it = m_inmap.begin(); it != m_inmap.end(); ++it) {
+    if (it->second.find(global_image_id) != it->second.end()) {
+      return it->first;
+    }
+  }
+
+  return UNMAPPED_INSTANCE_ID;
+}
+
+template <typename I>
+std::string Policy<I>::map(const std::string &global_image_id) {
+  assert(m_lock.is_wlocked());
+
+  std::string instance_id = do_map(global_image_id);
+  auto ins = m_inmap[instance_id].emplace(global_image_id);
+  assert(ins.second);
+
+  m_info.emplace(global_image_id, ImageInfo());
+  return instance_id;
+}
+
+template <typename I>
+bool Policy<I>::unmap(const std::string &global_image_id) {
+  assert(m_lock.is_wlocked());
+
+  std::string instance_id = lookup(global_image_id);
+  if (instance_id == UNMAPPED_INSTANCE_ID) {
+    return false;
+  }
+
+  m_inmap[instance_id].erase(global_image_id);
+  return true;
+}
+
+template <typename I>
+void Policy<I>::update_images_added(const std::string &mirror_uuid,
+                                    ImageIds &&image_ids, Images *images,
+                                    Context *on_finish) {
+  // images we just mapped -- once we start maintaining state
+  // for images, image map update request can selectively
+  // update on-disk map.
+  Images mapped;
+
+  for (auto const &image_id : image_ids) {
+    dout(20) << ": image_id=" << image_id << dendl;
+
+    std::string instance_id = lookup(image_id.global_id);
+    if (instance_id == UNMAPPED_INSTANCE_ID) {
+      instance_id = map(image_id.global_id);
```

Comment:
This policy should really be asynchronous based on a periodic timer callback. When the policy sees a bunch of changes, it can update the in-memory state and set a timer for the map/unmap in-bulk.  The timer periodic can be kept short (i.e. 1 second), but when we come online and have 3000 images, we don't want to do one-off mappings.